lib/internal/webstreams/readablestream.js

```diff
@@ -0,0 +1,2689 @@
+
+const {
+  kAborted,
+} = require('internal/abort_controller');
+
+const {
+  MessageChannel,
+} = require('internal/worker/io');
+
+const {
+  kDeserialize,
+  kTransfer,
+  kTransferList,
+  JSTransferable,
+} = require('internal/worker/js_transferable');
+
+const {
+  queueMicrotask,
+} = require('internal/process/task_queues');
+
+const {
+  ArrayBufferViewGetBuffer,
+  ArrayBufferViewGetByteLength,
+  ArrayBufferViewGetByteOffset,
+  ArrayBufferGetByteLength,
+  AsyncIterator,
+  copyArrayBuffer,
+  customInspect,
+  dequeueValue,
+  ensureIsPromise,
+  enqueueValueWithSize,
+  extractHighWaterMark,
+  extractSizeAlgorithm,
+  lazyTransfer,
+  internalExtend,
+  isBrandCheck,
+  resetQueue,
+  setPromiseHandled,
+  transferArrayBuffer,
+  nonOpCancel,
+  nonOpPull,
+  nonOpStart,
+  kType,
+  kState,
+} = require('internal/webstreams/util');
+
+const {
+  WritableStreamDefaultWriter,
+
+  isWritableStream,
+  isWritableStreamLocked,
+  isWritableStreamDefaultController,
+  isWritableStreamDefaultWriter,
+
+  writableStreamAbort,
+  writableStreamCloseQueuedOrInFlight,
+  writableStreamDefaultWriterCloseWithErrorPropagation,
+  writableStreamDefaultWriterRelease,
+  writableStreamDefaultWriterWrite,
+} = require('internal/webstreams/writablestream');
+
+const assert = require('internal/assert');
+
+const kCancel = Symbol('kCancel');
+const kClose = Symbol('kClose');
+const kChunk = Symbol('kChunk');
+const kError = Symbol('kError');
+const kPull = Symbol('kPull');
+
+/**
+ * @typedef {import('../abort_controller').AbortSignal} AbortSignal
+ * @typedef {import('./queuingstrategies').QueuingStrategy} QueuingStrategy
+ * @typedef {import('./queuingstrategies').QueuingStrategySize
+ * } QueuingStrategySize
+ * @typedef {import('./writablestream').WritableStream} WritableStream
+ *
+ * @typedef {ReadableStreamDefaultController | ReadableByteStreamController
+ * } ReadableStreamController
+ *
+ * @typedef {ReadableStreamDefaultReader | ReadableStreamBYOBReader
+ * } ReadableStreamReader
+ *
+ * @callback UnderlyingSourceStartCallback
+ * @param {ReadableStreamController} controller
+ * @returns { any | Promise<void> }
+ *
+ * @callback UnderlyingSourcePullCallback
+ * @param {ReadableStreamController} controller
+ * @returns { Promise<void> }
+ *
+ * @callback UnderlyingSourceCancelCallback
+ * @param {any} reason
+ * @returns { Promise<void> }
+ *
+ * @typedef {{
+ *   readable: ReadableStream,
+ *   writable: WritableStream,
+ * }} ReadableWritablePair
+ *
+ * @typedef {{
+ *   preventClose? : boolean,
+ *   preventAbort? : boolean,
+ *   preventCancel? : boolean,
+ *   signal? : AbortSignal,
+ * }} StreamPipeOptions
+ *
+ * @typedef {{
+ *   start? : UnderlyingSourceStartCallback,
+ *   pull? : UnderlyingSourcePullCallback,
+ *   cancel? : UnderlyingSourceCancelCallback,
+ *   type? : "bytes",
+ *   autoAllocateChunkSize? : number
+ * }} UnderlyingSource
+ *
+ */
+
+class ReadableStream extends JSTransferable {
```

Comment:
Unfortunately not without another hack that breaks other tests. We're a bit limited here

Reply:
To clarify, it's possible by dipping down into C++ land to create the transferable classes and use Object.defineOwnProperties to add the methods and accessors in JS land. It's not as elegant from a syntax and maintenance perspective but we can't clone/transfer objects unless the JS object is attached at the native layer to our BaseObject. The other way is to use the makeTransferable utility method we added recently but that breaks a couple of the other tests because it also messes with the prototype and constructor.

Reply:
Since this PR is passing CI and it's already pretty complicated as it is, I'll make the changes we need to get the IDL harness test completely passing in a separate follow-on PR